File: fsharp/FsSearch/FsSearch/FileTypes.fs
namespace FsSearch

open System
open System.Collections.Generic
open System.IO
open System.Text.Json
open System.Xml.Linq

type FileType = 
    | Unknown = 0
    | Archive = 1
    | Binary  = 2
    | Code    = 3
    | Text    = 4
    | Xml     = 5

type FileTypesDictionary = Dictionary<string, List<Dictionary<string,Object>>>

type FileTypes() =
    static let archive = "archive"
    static let binary = "binary"
    static let code = "code"
    static let searchable = "searchable"
    static let text = "text"
    static let xml = "xml"

    let ExtensionSet (s : string) =
        List.fold
            (fun (acc: Set<string>) (ext : string) -> Set.add ext acc)
            Set.empty
            (FileUtil.ExtensionsListFromString s)


    let PopulateFileTypesFromJson (jsonString : string) =
        let fileTypesDictionary = Dictionary<string, ISet<string>>()
        let filetypesDict = JsonSerializer.Deserialize<FileTypesDictionary>(jsonString)
        let filetypeDicts = filetypesDict.["filetypes"]
        for filetypeDict in filetypeDicts do
            let name = (filetypeDict.["type"] :?> JsonElement).GetString()
            let extensions =
                [ for x in (filetypeDict.["extensions"] :?> JsonElement).EnumerateArray() do
                    yield "." + x.GetString() ]
            fileTypesDictionary.Add(name, HashSet<String>(extensions))
        let allText = HashSet<String>(fileTypesDictionary.[text])
        allText.UnionWith(fileTypesDictionary.[code])
        allText.UnionWith(fileTypesDictionary.[xml])
        if fileTypesDictionary.Remove(text) then
            fileTypesDictionary.Add(text, allText)
        let searchableSet = HashSet<String>(fileTypesDictionary.[text])
        searchableSet.UnionWith(fileTypesDictionary.[binary])
        searchableSet.UnionWith(fileTypesDictionary.[archive])
        fileTypesDictionary.Add(searchable, searchableSet)
        fileTypesDictionary

    let PopulateFileTypesFromXml (xmlString : string) =
        let fileTypesDictionary = Dictionary<string, ISet<string>>()
        let filetypes = XDocument.Parse(xmlString).Descendants(XName.Get("filetype"))
        for f in filetypes do
            let name = [for a in f.Attributes(XName.Get("name")) do yield a.Value].Head
            let extSet =  ExtensionSet [for e in f.Descendants(XName.Get("extensions")) do yield e.Value].Head
            fileTypesDictionary.Add(name, HashSet<String>(extSet))
        let allText = HashSet<String>(fileTypesDictionary.[text])
        allText.UnionWith(fileTypesDictionary.[code])
        allText.UnionWith(fileTypesDictionary.[xml])
        if fileTypesDictionary.Remove(text) then
            fileTypesDictionary.Add(text, allText)
        let searchableSet = HashSet<String>(fileTypesDictionary.[text])
        searchableSet.UnionWith(fileTypesDictionary.[binary])
        searchableSet.UnionWith(fileTypesDictionary.[archive])
        fileTypesDictionary.Add(searchable, searchableSet)
        fileTypesDictionary

//    let _fileTypesResource = EmbeddedResource.GetResourceFileContents("FsSearch.Resources.filetypes.xml")
    let _fileTypesResource = EmbeddedResource.GetResourceFileContents("FsSearch.Resources.filetypes.json")
//    let _fileTypesDictionary = PopulateFileTypesFromXml(_fileTypesResource)
    let _fileTypesDictionary = PopulateFileTypesFromJson(_fileTypesResource)

    // read-only member properties
    member this.FileTypesDictionary = _fileTypesDictionary

    static member FromName (name : string) : FileType =
        let lname = name.ToLowerInvariant()
        if lname.Equals(text) then FileType.Text
        else if lname.Equals(binary) then FileType.Binary
        else if lname.Equals(archive) then FileType.Archive
        else if lname.Equals(code) then FileType.Code
        else if lname.Equals(xml) then FileType.Xml
        else FileType.Unknown

    static member ToName (fileType : FileType) : string =
        match fileType with
        | FileType.Archive -> "Archive"
        | FileType.Binary -> "Binary"
        | FileType.Code -> "Code"
        | FileType.Text -> "Text"
        | FileType.Xml -> "Xml"
        | _ -> "Unknown"

    member this.GetFileType (f : FileInfo) : FileType =
        if this.IsArchiveFile f then FileType.Archive
        else if this.IsBinaryFile f then FileType.Binary
        else if this.IsCodeFile f then FileType.Code
        else if this.IsXmlFile f then FileType.Xml
        else if this.IsTextFile f then FileType.Text
        else FileType.Unknown

    member this.IsArchiveFile (f : FileInfo) : bool =
        Seq.exists (fun x -> x = f.Extension.ToLowerInvariant()) this.FileTypesDictionary.[archive]

    member this.IsBinaryFile (f : FileInfo) : bool =
        Seq.exists (fun x -> x = f.Extension.ToLowerInvariant()) this.FileTypesDictionary.[binary]

    member this.IsCodeFile (f : FileInfo) : bool =
        Seq.exists (fun x -> x = f.Extension.ToLowerInvariant()) this.FileTypesDictionary.[code]

    member this.IsSearchableFile (f : FileInfo) : bool =
        Seq.exists (fun x -> x = f.Extension.ToLowerInvariant()) this.FileTypesDictionary.[searchable]

    member this.IsTextFile (f : FileInfo) : bool =
        Seq.exists (fun x -> x = f.Extension.ToLowerInvariant()) this.FileTypesDictionary.[text]

    member this.IsUnknownFile (f : FileInfo) : bool =
        not (this.IsSearchableFile f)

    member this.IsXmlFile (f : FileInfo) : bool =
        Seq.exists (fun x -> x = f.Extension.ToLowerInvariant()) this.FileTypesDictionary.[xml]
    ;;

File: csharp/CsSearch/CsSearchLib/FileTypes.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Xml.Linq;

using FileTypesDictionary = System.Collections.Generic.Dictionary<string, System.Collections.Generic.List<System.Collections.Generic.Dictionary<string,object>>>;

namespace CsSearchLib
{
	public enum FileType
	{
		Unknown,
		Archive,
		Binary,
		Code,
		Text,
		Xml
	};

	public class FileTypes
	{
		public readonly ISet<string> CurrentAndParentDirs = new HashSet<string> {".", ".."};

		private readonly string _fileTypesResource;
		private readonly IDictionary<string, ISet<string>> _fileTypesDictionary;

		private const string Archive = "archive";
		private const string Binary = "binary";
		private const string Code = "code";
		private const string Searchable = "searchable";
		private const string Text = "text";
		private const string Xml = "xml";

		public FileTypes()
		{
			// _fileTypesResource = EmbeddedResource.GetResourceFileContents("CsSearch.Resources.filetypes.xml");
			_fileTypesResource = EmbeddedResource.GetResourceFileContents("CsSearch.Resources.filetypes.json");
			_fileTypesDictionary = new Dictionary<string, ISet<string>>();
			// PopulateFileTypesFromXml();
			PopulateFileTypesFromJson();
		}

		private void PopulateFileTypesFromJson()
		{
			var filetypesDict = JsonSerializer.Deserialize<FileTypesDictionary>(_fileTypesResource);
			var filetypeDicts = filetypesDict["filetypes"];
			foreach (var filetypeDict in filetypeDicts)
			{
				var name = ((JsonElement)filetypeDict["type"]).GetString();
				var extensions = ((JsonElement)filetypeDict["extensions"]).EnumerateArray()
					.Select(x => "." + x.GetString());
				var extensionSet = new HashSet<string>(extensions);
				_fileTypesDictionary[name] = extensionSet;
			}
			_fileTypesDictionary[Text].UnionWith(_fileTypesDictionary[Code]);
			_fileTypesDictionary[Text].UnionWith(_fileTypesDictionary[Xml]);
			_fileTypesDictionary[Searchable] = new HashSet<string>(_fileTypesDictionary[Text]);
			_fileTypesDictionary[Searchable].UnionWith(_fileTypesDictionary[Binary]);
			_fileTypesDictionary[Searchable].UnionWith(_fileTypesDictionary[Archive]);
		}

		private void PopulateFileTypesFromXml()
		{
			var doc = XDocument.Parse(_fileTypesResource);
			foreach (var f in doc.Descendants("filetype"))
			{
				var name = f.Attributes("name").First().Value;
				var extensions = f.Descendants("extensions").First().Value;
				var extensionSet = new HashSet<string>(extensions.Split(new[]{' ', '\n'}).Select(x => "." + x));
				_fileTypesDictionary[name] = extensionSet;
			}
			_fileTypesDictionary[Text].UnionWith(_fileTypesDictionary[Code]);
			_fileTypesDictionary[Text].UnionWith(_fileTypesDictionary[Xml]);
			_fileTypesDictionary[Searchable] = new HashSet<string>(_fileTypesDictionary[Text]);
			_fileTypesDictionary[Searchable].UnionWith(_fileTypesDictionary[Binary]);
			_fileTypesDictionary[Searchable].UnionWith(_fileTypesDictionary[Archive]);
		}

		public static FileType FromName(string name)
		{
			return string.IsNullOrEmpty(name)
				? FileType.Unknown
				: name.ToLowerInvariant() switch
			{
				Archive => FileType.Archive,
				Binary => FileType.Binary,
				Code => FileType.Code,
				Text => FileType.Text,
				Xml => FileType.Xml,
				_ => FileType.Unknown
			};
		}

		public FileType GetFileType(FileInfo f)
		{
			if (IsArchiveFile(f)) return FileType.Archive;
			if (IsBinaryFile(f)) return FileType.Binary;
			if (IsCodeFile(f)) return FileType.Code;
			if (IsXmlFile(f)) return FileType.Xml;
			return IsTextFile(f) ? FileType.Text : FileType.Unknown;
		}

		public bool IsArchiveFile(FileInfo f)
		{
			return _fileTypesDictionary[Archive].Contains(f.Extension.ToLowerInvariant());
		}

		public bool IsBinaryFile(FileInfo f)
		{
			return _fileTypesDictionary[Binary].Contains(f.Extension.ToLowerInvariant());
		}

		public bool IsCodeFile(FileInfo f)
		{
			return _fileTypesDictionary[Code].Contains(f.Extension.ToLowerInvariant());
		}

		public bool IsSearchableFile(FileInfo f)
		{
			return _fileTypesDictionary[Searchable].Contains(f.Extension.ToLowerInvariant());
		}

		public bool IsTextFile(FileInfo f)
		{
			return _fileTypesDictionary[Text].Contains(f.Extension.ToLowerInvariant());
		}

		public bool IsUnknownFile(FileInfo f)
		{
			return !IsSearchableFile(f);
		}

		public bool IsXmlFile(FileInfo f)
		{
			return _fileTypesDictionary[Xml].Contains(f.Extension.ToLowerInvariant());
		}
	}
}
